### PR TITLE
feat: storage disaster recovery

### PR DESCRIPTION
**Pull request by iliapolo**
Prepare for and offer mitigation for a storage disaster.

### Implementation Notes

The PR includes a lot of files changed but its mostly simple refactoring. The idea is that instead of creating buckets with `new Bucket`, we now go though a factory that will:

1. Create a failover bucket for every bucket.
2. Decide if the failover or the original bucket should be used in the current deployment. 

----

*By submitting this pull request, I confirm that my contribution is made under the terms of the Apache-2.0 license*

### CODE REVIEW

File: docs/operator-runbook.md [hunk 1]
```diff
@@ -12,6 +12,54 @@ readily available.
...
+And deploy this to your environment. This will swap out all the original buckets with the pre-populated failover buckets.
+Note that any data that was indexed in construct hub post the creation of the snapshot, will not be available immediately once you perform the failover.
+Construct hub will pick up discovery from the marker that was included in the last snapshot.
+
+When you restore the original data and are ready to go back to the original buckets, simply remove this property and deploy again.
+
+#### When NOT to use this procedure
+
+If the data loss/corruption is self-inflicted and continuous, i.e construct hub misbehaves and mutates its own data in a faulty manner.
+In this case switching to the failover won't help because the bad behavior will be applied on the failover buckets.
```
Comment (iliapolo) [inline docs/operator-runbook.md:56]: We could also make it so switching to the failover buckets will disable the entire pipeline and make construct hub static. I haven't had time to look into that just yet.

File: src/__tests__/backend/deny-list/__snapshots__/deny-list.test.ts.snap [hunk 2]
```diff
@@ -3,6 +3,24 @@
 exports[`defaults - empty deny list 1`] = `
 Object {
   "Outputs": Object {
+    "DenyListSnapshotCommand31529C83": Object {
+      "Description": "Snapshot Default/DenyList/Bucket",
+      "Value": Object {
+        "Fn::Join": Array [
+          "",
+          Array [
+            "aws s3 sync s3://",
```
Comment (eladb) [inline src/__tests__/backend/deny-list/__snapshots__/deny-list.test.ts.snap:12]: Instead of a bunch of outputs which users are expected to copy & paste, can't we just write a simple lambda function or a CodeBuild job that takes a full snapshot?

File: src/construct-hub.ts [hunk 3]
```diff
@@ -150,6 +151,17 @@ export interface ConstructHubProps {
...
+  /**
+   * Wire construct hub to use the failover storage buckets.
+   *
+   * Do not activate this property until you've populated your failover buckets
+   * with the necessary data.
+   *
+   * @see https://github.com/cdklabs/construct-hub/blob/dev/docs/operator-runbook.md#storage-disaster
+   * @default false
+   */
+  readonly failoverStorageActive?: boolean;
```
Comment (eladb) [inline src/construct-hub.ts:163]: The word "active" here is confusing (to me). I think this should be sufficient:

```suggestion
  readonly failoverStorage?: boolean;
```

File: src/backend/deny-list/deny-list.ts [hunk 4]
```diff
@@ -57,6 +58,13 @@ export interface DenyListProps {
    * The monitoring system.
    */
   readonly monitoring: Monitoring;
+
+  /**
+   * Factory for creating storage on S3.
+   *
+   * @default - The default storage factory (plain s3 bucket).
+   */
+  readonly storageFactory?: S3StorageFactory;
```
Comment (eladb) [inline src/backend/deny-list/deny-list.ts:67]: Interesting approach (did I hear someone say "cool construct library"?)

File: src/package-sources/code-artifact.ts [hunk 5]
```diff
@@ -4,16 +4,25 @@ import { Rule } from '@aws-cdk/aws-events';
...
 import { METRICS_NAMESPACE, MetricName, DOMAIN_NAME_DIMENSION, DOMAIN_OWNER_DIMENSION, REPOSITORY_NAME_DIMENSION } from './codeartifact/constants.lambda-shared';
 
 export interface CodeArtifactProps {
+
+  /**
+   * Factory for creating storage on S3.
+   *
+   * @default - The default storage factory (plain s3 bucket).
+   */
+  readonly storageFactory?: S3StorageFactory;
```
Comment (eladb) [inline src/package-sources/code-artifact.ts:24]: Why optional?

```suggestion
  readonly storageFactory: S3StorageFactory;
```

File: src/package-sources/npmjs.ts [hunk 6]
```diff
@@ -49,21 +50,29 @@ export interface NpmJsProps {
    * @default Duration.minutes(5)
    */
   readonly canarySla?: Duration;
+
+  /**
+   * Factory for creating storage on S3.
+   *
+   * @default - The default storage factory (plain s3 bucket).
+   */
+  readonly storageFactory?: S3StorageFactory;
```
Comment (eladb) [inline src/package-sources/npmjs.ts:59]: Why optional?

File: src/package-sources/npmjs.ts [hunk 7]
```diff
@@ -445,7 +454,7 @@ export class NpmJs implements IPackageSource {
       treatMissingData: TreatMissingData.BREACHING,
       threshold: visibilitySla.toSeconds(),
     });
-    monitoring.addLowSeverityAlarm('New version visibility SLA breached', alarm);
+    monitoring.addHighSeverityAlarm('New version visibility SLA breached', alarm);
```
Comment (eladb) [inline src/package-sources/npmjs.ts:457]: Is this related to this change? Might be worth splitting to a separate PR

File: src/package-sources/npmjs/npm-js-follower.lambda.ts [hunk 8]
```diff
@@ -204,7 +204,7 @@ async function loadLastTransactionMarker(
     const dbUpdateSeq = (await registry.info()).update_seq;
     if (dbUpdateSeq < data.marker) {
       console.warn(`Current DB update_seq (${dbUpdateSeq}) is lower than marker (CouchDB instance was likely replaced), resetting to 0!`);
-      return { marker: '0', knownVersions: data.knownVersion };
+      return { marker: '6121592', knownVersions: data.knownVersion };
```
Comment (eladb) [inline src/package-sources/npmjs/npm-js-follower.lambda.ts:207]: ?

File: src/s3/storage.ts [hunk 9]
```diff
@@ -0,0 +1,45 @@
...
+ */
+export class S3StorageFactory {
+
+  private failoverActive: boolean;
+
+  constructor(props: S3StorageFactoryProps = {}) {
+    this.failoverActive = props.failoverActive ?? false;
+  };
+
+  public newBucket(scope: Construct, id: string, props?: s3.BucketProps): s3.Bucket {
```
Comment (eladb) [inline src/s3/storage.ts:22]: docstrings

File: src/webapp/index.ts [hunk 10]
```diff
@@ -131,6 +132,13 @@ export interface WebAppProps extends WebappConfigProps {
    */
   readonly monitoring: Monitoring;
 
+  /**
+   * Factory for creating storage on S3.
+   *
+   * @default - The default storage factory (plain s3 bucket).
+   */
+  readonly storageFactory?: S3StorageFactory;
```
Comment (eladb) [inline src/webapp/index.ts:140]: Why optional?

File: src/s3/storage.ts [hunk 11]
```diff
@@ -0,0 +1,45 @@
...
+ * Properties for `S3StorageFactory`
+ */
+export interface S3StorageFactoryProps {
+  readonly failoverActive?: boolean;
+}
+
+/**
+ * Create s3 storage resources.
+ */
+export class S3StorageFactory {
```
Comment (eladb) [inline src/s3/storage.ts:14]: Please add some unit and integ tests for this

**Review by eladb — CHANGES_REQUESTED** [review]
I like the approach (so nice what you could do with software :-)).

File: src/backend/deny-list/deny-list.ts [hunk 12]
```diff
@@ -87,11 +88,12 @@ export class DenyList extends Construct implements IDenyList {
...
     super(scope, id);
 
-    this.bucket = new s3.Bucket(this, 'Bucket', {
+    const storageFactory = S3StorageFactory.getOrCreate(this);
+
+    this.bucket = storageFactory.newBucket(this, 'Bucket', {
       blockPublicAccess: s3.BlockPublicAccess.BLOCK_ALL,
       encryption: s3.BucketEncryption.S3_MANAGED,
       enforceSSL: true,
-      removalPolicy: RemovalPolicy.DESTROY,
```
Comment (iliapolo) [inline src/backend/deny-list/deny-list.ts:94]: This bucket cannot be destroyed at the moment since its versioned, and our deletion handler doesn't delete object versions.

File: src/backend/license-list/license-list.ts [hunk 13]
```diff
@@ -36,18 +37,17 @@ export class LicenseList extends Construct implements ILicenseList {
   public constructor(scope: Construct, id: string, props: LicenseListProps) {
     super(scope, id);
 
-    this.bucket = new s3.Bucket(this, 'Bucket', {
+    const storageFactory = S3StorageFactory.getOrCreate(this);
+    this.bucket = storageFactory.newBucket(this, 'Bucket', {
       blockPublicAccess: s3.BlockPublicAccess.BLOCK_ALL,
       encryption: s3.BucketEncryption.S3_MANAGED,
       enforceSSL: true,
-      removalPolicy: RemovalPolicy.DESTROY,
```
Comment (iliapolo) [inline src/backend/license-list/license-list.ts:43]: This bucket cannot be destroyed at the moment since its versioned, and our deletion handler doesn't delete object versions.

File: src/package-sources/npmjs.ts (hunk 7 above)
Comment (iliapolo) [reply src/package-sources/npmjs.ts:450]: No, not sure how this got here. I'll remove it.

File: src/package-sources/npmjs/npm-js-follower.lambda.ts (hunk 8 above)
Comment (iliapolo) [reply src/package-sources/npmjs/npm-js-follower.lambda.ts:207]: Phew, good catch!

File: src/__tests__/backend/deny-list/__snapshots__/deny-list.test.ts.snap (hunk 2 above)
Comment (iliapolo) [reply src/__tests__/backend/deny-list/__snapshots__/deny-list.test.ts.snap:12]: Yes, this will be phase two

File: docs/operator-runbook.md [hunk 14]
```diff
@@ -12,6 +12,56 @@ readily available.
...
+
+After construct hub deployment finishes, at the time of your choosing, locate those outputs in the CloudFormation console, under the "Outputs" tab, and run the commands:
+
+```console
+aws s3 sync s3://construct-hub-dev-constructhubdenylistbucket1b3c2-1n2loz0z3u2x4 s3://construct-hub-dev-constructhubdenylistfailoverbuc-hnxtur2i9smy
+aws s3 sync s3://construct-hub-dev-constructhubingestionconfigbuck-lrjjt5rdbfnr s3://construct-hub-dev-constructhubingestionfailoverco-flcmkc25ra5e
+aws s3 sync s3://construct-hub-dev-constructhublicenselistbucket93-1bdf207eys9et s3://construct-hub-dev-constructhublicenselistfailover-1jqfoba4mxfny
+aws s3 sync s3://construct-hub-dev-constructhubpackagedatadc5ef35e-4lgmx5xe9nv2 s3://construct-hub-dev-constructhubfailoverpackagedata-qr2qge4kodvv
+aws s3 sync s3://construct-hub-dev-constructhubsourcesnpmjsstaging-14eppmcw5vsr3 s3://construct-hub-dev-constructhubsourcesfailovernpmj-zc2sciapis06
+aws s3 sync s3://construct-hub-dev-constructhubwebappwebsitebucket-49110nn5q2ro s3://construct-hub-dev-constructhubwebappfailoverwebsi-1o2u8fix6ujdi
```
Comment (RomainMuller) [inline docs/operator-runbook.md:34]: You might want to replace the bucket names with placeholders here.

File: src/__tests__/backend/deny-list/integ/deny-list.integ.cdkout/TestDenyList.template.json [hunk 15]
```diff
@@ -841,6 +841,90 @@
         "aws:cdk:path": "TestDenyList/DenyList/Bucket/AllowBucketNotificationsToTestDenyListPrunePruneHandlerDB8102B7"
       }
     },
+    "DenyListFailoverBucketC0405145": {
```
Comment (RomainMuller) [inline src/__tests__/backend/deny-list/integ/deny-list.integ.cdkout/TestDenyList.template.json:844]: Arguably the buckets that are only here to hold static configuration objects that are injected by the deployment... probably don't need a failover.

Comment (iliapolo) [reply src/__tests__/backend/deny-list/integ/deny-list.integ.cdkout/TestDenyList.template.json:844]: That is true, decided not to do this "optimization" - doesn't really hurt

File: docs/operator-runbook.md (hunk 14 above)
Comment (iliapolo) [reply docs/operator-runbook.md:34]: 👍

File: src/backend/deny-list/deny-list.ts (hunk 12 above)
Comment (eladb) [reply src/backend/deny-list/deny-list.ts:94]: Make the removal policy explicit RETAIN

File: src/backend/deny-list/deny-list.ts [hunk 16]
```diff
@@ -101,7 +103,6 @@ export class DenyList extends Construct implements IDenyList {
     const upload = new s3deploy.BucketDeployment(this, 'BucketDeployment', {
       destinationBucket: this.bucket,
       prune: true,
-      retainOnDelete: false,
```
Comment (eladb) [inline src/backend/deny-list/deny-list.ts:104]: Make explicit true

File: src/s3/storage.ts [hunk 17]
```diff
@@ -0,0 +1,72 @@
...
+ * Create s3 storage resources.
+ */
+export class S3StorageFactory extends Construct {
+
+  /**
+   * Retrieve or create the storage factory for the current scope.
+   *
+   * This is stack singleton.
+   */
+  public static getOrCreate(scope: Construct, props: S3StorageFactoryProps = {}): S3StorageFactory {
```
Comment (eladb) [inline src/s3/storage.ts:26]: Style: Normally we call this `of()`

**Review by eladb — APPROVED** [review]

**Comment by eladb** [comment]
Minor comments. Added do not merge

File: src/backend/deny-list/deny-list.ts (hunk 12 above)
Comment (iliapolo) [reply src/backend/deny-list/deny-list.ts:94]: Done
